iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerWithEngine.java

```diff
@@ -2271,6 +2274,116 @@ public void testStatWithPartitionedCTAS() {
     checkColStat("target", "dept");
   }
 
+  @Test
+  public void testAsOfTimestamp() throws IOException, InterruptedException {
+    Table table = prepareTableWithVersions(2);
+
+    List<Object[]> rows = shell.executeStatement(
+        "SELECT * FROM customers FOR SYSTEM_TIME AS OF '" + timestampAfterSnapshot(table, 0) + "'");
+
+    Assert.assertEquals(3, rows.size());
+
+    rows = shell.executeStatement(
+        "SELECT * FROM customers FOR SYSTEM_TIME AS OF '" + timestampAfterSnapshot(table, 1) + "'");
+
+    Assert.assertEquals(4, rows.size());
+
+    AssertHelpers.assertThrows("should throw exception", IllegalArgumentException.class,
+        "Cannot find a snapshot older than 1970-01-01 00:00:00", () -> {
+          shell.executeStatement("SELECT * FROM customers FOR SYSTEM_TIME AS OF '1970-01-01 00:00:00'");
+        });
+  }
+
+  @Test
+  public void testAsOfVersion() throws IOException, InterruptedException {
+    Table table = prepareTableWithVersions(2);
+
+    HistoryEntry first = table.history().get(0);
+    List<Object[]> rows =
+        shell.executeStatement("SELECT * FROM customers FOR SYSTEM_VERSION AS OF " + first.snapshotId());
+
+    Assert.assertEquals(3, rows.size());
+
+    HistoryEntry second = table.history().get(1);
+    rows = shell.executeStatement("SELECT * FROM customers FOR SYSTEM_VERSION AS OF " + second.snapshotId());
+
+    Assert.assertEquals(4, rows.size());
+
+    AssertHelpers.assertThrows("should throw exception", IllegalArgumentException.class,
+        "Cannot find snapshot with ID 1234", () -> {
+          shell.executeStatement("SELECT * FROM customers FOR SYSTEM_VERSION AS OF 1234");
+        });
+  }
+
+  @Test
+  public void testAsOfTimestampWithJoins() throws IOException, InterruptedException {
+    Table table = prepareTableWithVersions(4);
+
+    List<Object[]> rows = shell.executeStatement("SELECT * FROM " +
+        "customers FOR SYSTEM_TIME AS OF '" + timestampAfterSnapshot(table, 0) + "' fv, " +
+        "customers FOR SYSTEM_TIME AS OF '" + timestampAfterSnapshot(table, 1) + "' sv " +
+        "WHERE fv.first_name=sv.first_name");
+
+    Assert.assertEquals(4, rows.size());
+
+    rows = shell.executeStatement("SELECT * FROM " +
+         "customers FOR SYSTEM_TIME AS OF '" + timestampAfterSnapshot(table, 1) + "' sv, " +
+         "customers FOR SYSTEM_TIME AS OF '" + timestampAfterSnapshot(table, 2) + "' tv " +
+         "WHERE sv.first_name=tv.first_name");
+
+    Assert.assertEquals(8, rows.size());
+
+    rows = shell.executeStatement("SELECT * FROM " +
+        "customers FOR SYSTEM_TIME AS OF '" + timestampAfterSnapshot(table, 2) + "' sv, " +
+        "customers lv " +
+        "WHERE sv.first_name=lv.first_name");
+
+    Assert.assertEquals(14, rows.size());
+  }
+
+  /**
+   * Creates the 'customers' table with the default records and creates extra snapshots by inserting one more line
+   * into the table.
+   * @param versions The number of snapshots we want to create
```

Comment:
Create table does not create history element, and the first insert creates a new version. That is why the loop runs with like this (notice: `versions - 1`):
```
for (int i = 0; i < versions - 1; ++i) {
```

Reply:
Yeah, I think there's a bit of a difference between snapshots and history. If you create an empty table, it will have 1 snapshot, but 0 history - is that right? If that's true, maybe we could reword the javadoc `The number of history elements we want to create`? (and the method name)

Reply:
Done, Thanks!